sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/multi/MultiServiceEventClient.java

```diff
@@ -0,0 +1,402 @@
+        }
+
+        // Iterate over offersByService and find out if the services in question still want the resources.
+        // Any unwanted resources then get added to unexpectedResources.
+        boolean anyFailedClients = false;
+        for (Map.Entry<String, Map<Protos.OfferID, OfferResources>> entry : offersByService.entrySet()) {
+            String serviceName = entry.getKey();
+            Collection<OfferResources> serviceOffers = entry.getValue().values();
+
+            Optional<AbstractScheduler> service = multiServiceManager.getService(serviceName);
+            if (!service.isPresent()) {
+                // (CASE 2) Old or invalid service name. Consider all resources for this service as unexpected.
+                LOGGER.info("  {} cleanup result: unknown service, all resources unexpected", serviceName);
+                for (OfferResources serviceOffer : serviceOffers) {
+                    getEntry(unexpectedResources, serviceOffer.getOffer()).addAll(serviceOffer.getResources());
+                }
+            } else {
+                // Construct offers containing (only) these resources and pass them to the service.
+                // See which of the resources are now unexpected by the service.
+                List<Protos.Offer> offersToSend = new ArrayList<>(serviceOffers.size());
+                for (OfferResources serviceOfferResources : serviceOffers) {
+                    offersToSend.add(serviceOfferResources.getOffer().toBuilder()
+                            .clearResources()
+                            .addAllResources(serviceOfferResources.getResources())
+                            .build());
+                }
+                // (CASE 3) The service has returned the subset of these resources which are unexpected.
+                // Add those to unexpectedResources.
+                // Note: We're careful to only invoke this once per service, as the call is likely to be expensive.
+                UnexpectedResourcesResponse response = service.get().getUnexpectedResources(offersToSend);
+                LOGGER.info("  {} cleanup result: {} with {} unexpected resources in {} offer{}",
+                        serviceName,
+                        response.result,
+                        response.offerResources.stream()
+                                .collect(Collectors.summingInt(or -> or.getResources().size())),
+                        response.offerResources.size(),
+                        response.offerResources.size() == 1 ? "" : "s");
+                switch (response.result) {
+                case FAILED:
+                    // We should be able to safely skip this service and proceed to the next one. For this round,
+                    // the service just won't have anything added to unexpectedResources. We play it safe by telling
+                    // upstream to do a short decline for the unprocessed resources.
```

Comment:
This statement confuses me.  You can only accept/decline offers as a whole.  A short decline of a subset of resources is not a thing as far as I know.

Reply:
Cleaned this up in a follow-up PR.